src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -556,6 +556,227 @@ inline bool ReduceAxesOpForwardStorage(const nnvm::NodeAttrs& attrs,
+            const int nWorkers, void *pIdxStorage, const idxStorageType storageType,
+            const TBlob& input, const TBlob& output) {
+  using namespace mxnet_op;
+  const auto pIn = input.dptr<DType>();
+  auto *pOut = output.dptr<DType>();
+  const auto nSize = shape[axis];
+  const auto num_threads = shape.Size() / nSize;
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  Kernel<argmax, xpu>::Launch(s, num_threads * nWorkers, nWorkers, pIn, pOut, nSize,
+                              step, shift, pIdxStorage, storageType);
+  if (nWorkers > 1) {
+    Kernel<argmax_reduce, xpu>::Launch(s, num_threads, nWorkers, pIn, pOut,
+                                       step, shift, pIdxStorage, storageType);
+  }
+}
+
+template<typename xpu>
+int DefineNumbWorkers(const TShape &shape, int axis);
+
+template<typename xpu>
+void ArgMax(const nnvm::NodeAttrs& attrs,
+            const OpContext& ctx,
+            const std::vector<TBlob>& inputs,
+            const std::vector<OpReqType>& req,
+            const std::vector<TBlob>& outputs) {
+  const ReduceAxisParam& param = nnvm::get<ReduceAxisParam>(attrs.parsed);
+  auto shape = inputs[0].shape_;
+  size_t shift = shape.Size();     // to reading data starting of each thread
+  CHECK_NE(shift, 0U) << "attempt to search an empty sequence";
+
+  size_t step = 1;  // step over the input array which will be used by each thread in kernel
+  int axis;
+  if (param.axis.has_value()) {
+    axis = CheckAxis(param.axis.value(), shape.ndim());
+    // cannot do argmax in an empty dimension
+    CHECK_NE(shape[axis], 0)
+      << "searching input tensor of shape " << shape
+      << " along axis = " << axis << " of zero dim-size is not allowed";
+
+    if (shape.ndim() == 1)
+      shape = AxisShapeCompact(shape, &axis, true);
+
+    // Calculate step & shift
+    auto i = shape.ndim();
+    while (--i > axis)
+      step *= shape[i];
+
+    shift = i? step*shape[i] : 1;
+  } else {
+    // If global reduction, reshape the input tensor into 2D shape (1, inputs[0].shape_.Size())
+    // and search on axis = 1.
+    shape = mxnet::TShape(2, 1);
+    shape[1] = shift;
+    axis = 1;
+  }
+
+  void *pIdxMemory = nullptr;
+  auto nWorkers = DefineNumbWorkers<xpu>(shape, axis);
+  idxStorageType storageType = undefined_st;
+  while (nWorkers > 1) {
+    const size_t num_items = nWorkers * shape.Size() / shape[axis];
+#if MXNET_USE_INT64_TENSOR_SIZE
+    pIdxMemory = AllocateDTypeMemory<xpu, index_t>(ctx, num_items);
+    if (pIdxMemory) {
+      storageType = int64_st;
+      break;
+    }
+#endif
+    if (shape[axis] <= UINT32_MAX) {
+      pIdxMemory = AllocateDTypeMemory<xpu, uint32_t>(ctx, num_items);
+      if (pIdxMemory) {
+        storageType = uint32_st;
+        break;
+      }
+    }
+    // Check if indexes can be stored in uint16 format
+    if (shape[axis] <= UINT16_MAX) {
+      pIdxMemory = AllocateDTypeMemory<xpu, uint16_t>(ctx, num_items);
```

Comment:
Is this working? I thought the index inside tensor object always use int32 or int64. Would this cause overflow if uint16_t is passed in. Also based on Google C++ style guide, we should avoid using unsigned integer to represent data type: https://google.github.io/styleguide/cppguide.html#Integer_Types

Reply:
No, this will never cause a problem because before allocation of pointer of `int64_t/uint32_t/uint16_t` types, we always check if it is possible to store indexes in corresponding formats. 
For instance
```
if (shape[axis] <= UINT16_MAX)
 pIdxMemory = AllocateDTypeMemory<xpu, uint16_t>(ctx, num_items);
```
The index will never be bigger than the size of corresponding dimension `shape[axis]`. 
Therefore, we should no overflow for any type of `index_t`.

Regarding the Google C ++ style guide, about not-using an unsigned integers, I'm not sure if this is a universal tip, especially when we are talking about a significant difference in runtime (and 5-10% I would consider a significant difference).

Reply:
> No, this will never cause a problem because before allocation of pointer of `int64_t/uint32_t/uint16_t` types, we always check if it is possible to store indexes in corresponding formats.
> For instance
> 
> ```
> if (shape[axis] <= UINT16_MAX)
>  pIdxMemory = AllocateDTypeMemory<xpu, uint16_t>(ctx, num_items);
> ```
> 
> The index will never be bigger than the size of corresponding dimension `shape[axis]`.
> Therefore, we should no overflow for any type of `index_t`.
> 
> Regarding the Google C ++ style guide, about not-using an unsigned integers, I'm not sure if this is a universal tip, especially when we are talking about a significant difference in run time (and 5-10% degradation which I observed, I would consider a significant difference).

Interesting. So by changing from uint32_t to int32_t, there is a 5~10% runtime degradation?

Reply:
@apeforest: I got email with your comments
```
Interesting. So by changing from uint32_t to int32_t, there is a 5~10% runtime degradation?
```
For some reasons I don't see it here... Anyway, I will respond on this one.
It's what you could found in https://github.com/apache/incubator-mxnet/blob/master/make/osx.mk : 
```
# Use int64_t type to represent the total number of elements in a tensor
# This will cause performance degradation reported in issue #14496
# Set to 1 for large tensor with tensor size greater than INT32_MAX i.e. 2147483647
# Note: the size of each dimension is still bounded by INT32_MAX
USE_INT64_TENSOR_SIZE = 0
```

Here is the link to that issue: https://github.com/apache/incubator-mxnet/issues/14496
The degradation here is also approx 7%

Reply:
I added this compiler flag in the make file. The performance degradation is from int32_t to int64_t.

My earlier comment to your PR is choosing int32_t over uint32_t and I was curious to know if there is any performance difference between int32_t and uint32_t

Reply:
I saw approx. 5% degradation when I was using `uint32_t`.I'm not sure, maybe it's just some kind of noise, and for such a conclusion, additional testing is necessary.
To make the code more robust, I decided not to use `unsigned` indexes (see https://github.com/apache/incubator-mxnet/pull/16218/commits/bf843d55de98d04fe19ee20703cd33b2912f7ef1 and https://github.com/apache/incubator-mxnet/pull/16218/commits/c7bf38d4e4ca481aa5004c446407cab38f1b70f1)